File: clj-advoc/src/clj_advoc/day_5.clj
(ns clj-advoc.day-5
  (:require [clojure.java.io :as io]
            [clojure.string :as s])
  (:import [java.lang Math]))

(defonce rowcol-sep 7)
(defonce conv-map {"f" "0"
                   "b" "1"
                   "r" "1"
                   "l" "0"})


(defn problem-input
  [filename]
  (let [rdr (io/reader (io/resource filename))]
    (reduce (fn [input line]
              (conj input line))
            []
            (line-seq rdr))))


(defn code->bin
  "Convert column code to a binary string."
  [code]
  (reduce (fn [string [match replacement]]
            (s/replace string match replacement))
          (s/lower-case code)
          conv-map))


(defn code->seat-id
  [code]
  (let [bin-code (code->bin code)
        [row col] (->> bin-code
                       (split-at rowcol-sep)
                       (map s/join)
                       (map #(Integer/parseInt % 2)))]
    (+ col (* 8 row))))


(defn solution-part-1
  [input]
  (let [ids (map code->seat-id input)]
    (apply max ids)))


(defn solution-part-2
  [input]
  (let [ids (map code->seat-id input)
        sort-ids (sort ids)]
    (loop [prev (first sort-ids)
           cur (second sort-ids)
           rest-ids (drop 2 sort-ids)]
      (cond
        (not= (inc prev) cur) (inc prev)
        (not (seq rest-ids)) -1
        :else (recur cur (first rest-ids) (rest rest-ids))))))


(defn print-solution
  [solution]
  (println "Answer: " solution))


(defn main
  []
  (let [inpt (problem-input "day-5-part-1.input")]
    (-> inpt
        (solution-part-1)
        (print-solution))
    (-> inpt
        (solution-part-2)
        (print-solution))))

File: py-advoc/day_5.py
import os

INPUT_PATH = "../inputs/"
TRANS_TABLE = str.maketrans('fbrl', '0110')


def problem_input(filename):
    path = os.path.join(INPUT_PATH, filename)
    inputs = []
    with open(path, 'r') as f:
        for n in f:
            inputs.append(n)

    return inputs


def get_seat_id(code):
    """
    For a given `code` return the corresponding seat id. The logic for
    this is as follows:
    The first 7 chars represent binary row (R)
    The remaining 3 chars represent binary column (C)
    Seat is defined as (R * 8) + C
    """
    bin_code = code.lower().translate(TRANS_TABLE)
    row, col = int(bin_code[:7], 2), int(bin_code[7:], 2)
    return (row * 8) + col


def solution_part_1(inpt):
    return max(map(get_seat_id, inpt))


def solution_part_2(inpt):
    ids = list(map(get_seat_id, inpt))
    ids.sort()
    prev = ids[0]
    for cur in ids[1:]:
        if (prev + 1) != cur:
            return prev + 1
        prev = cur
    return -1


if __name__ == "__main__":
    inpt = problem_input("day-5-part-1.input")
    ans = solution_part_1(inpt)
    print(f'Answer: {ans}')

    ans = solution_part_2(inpt)
    print(f'Answer: {ans}')
